clients/java/signalr/src/test/java/com/microsoft/aspnet/signalr/HubConnectionTest.java

```diff
@@ -304,12 +306,141 @@ public void registeringMultipleHandlersThatTakeParamsAndBothGetTriggered() throw
         hubConnection.start();
         mockTransport.receiveMessage("{}" + RECORD_SEPARATOR);
         mockTransport.receiveMessage("{\"type\":1,\"target\":\"add\",\"arguments\":[12]}" + RECORD_SEPARATOR);
-        hubConnection.send("add", 12);
 
         // Confirming that our handler was called and the correct message was passed in.
         assertEquals(24, value.get(), 0);
     }
 
+    @Test
+    public void invokeWaitsForCompletionMessage() throws Exception {
+        MockTransport mockTransport = new MockTransport();
+        HubConnection hubConnection = new HubConnection("http://example.com", mockTransport, true);
+
+        hubConnection.start();
+        mockTransport.receiveMessage("{}" + RECORD_SEPARATOR);
+
+        CompletableFuture<Integer> result = hubConnection.invoke(Integer.class, "echo", "message");
+        assertEquals("{\"type\":1,\"invocationId\":\"1\",\"target\":\"echo\",\"arguments\":[\"message\"]}" + RECORD_SEPARATOR, mockTransport.sentMessages.get(1));
+        assertFalse(result.isDone());
+
+        mockTransport.receiveMessage("{\"type\":3,\"invocationId\":\"1\",\"result\":42}" + RECORD_SEPARATOR);
+        assertTrue(result.isDone());
+
+        assertEquals(42, result.get(), 0);
+    }
+
+    @Test
+    public void multipleInvokesWaitForOwnCompletionMessage() throws Exception {
+        MockTransport mockTransport = new MockTransport();
+        HubConnection hubConnection = new HubConnection("http://example.com", mockTransport, true);
+
+        hubConnection.start();
+        mockTransport.receiveMessage("{}" + RECORD_SEPARATOR);
+
+        CompletableFuture<Integer> result = hubConnection.invoke(Integer.class, "echo", "message");
+        CompletableFuture<String> result2 = hubConnection.invoke(String.class, "echo", "message");
+        assertEquals("{\"type\":1,\"invocationId\":\"1\",\"target\":\"echo\",\"arguments\":[\"message\"]}" + RECORD_SEPARATOR, mockTransport.sentMessages.get(1));
+        assertEquals("{\"type\":1,\"invocationId\":\"2\",\"target\":\"echo\",\"arguments\":[\"message\"]}" + RECORD_SEPARATOR, mockTransport.sentMessages.get(2));
+        assertFalse(result.isDone());
+        assertFalse(result2.isDone());
+
+        mockTransport.receiveMessage("{\"type\":3,\"invocationId\":\"2\",\"result\":\"message\"}" + RECORD_SEPARATOR);
+        assertTrue(result2.isDone());
+        assertFalse(result.isDone());
+
+        mockTransport.receiveMessage("{\"type\":3,\"invocationId\":\"1\",\"result\":42}" + RECORD_SEPARATOR);
+        assertTrue(result.isDone());
+
+        assertEquals(42, result.get(), 0);
+        assertEquals("message", result2.get());
+    }
+
+    @Test
+    public void completionMessageValuesCanBeOutOfOrder() throws Exception {
```

Comment:
nit: technically this is a JsonHubProtocol test. Do we have those already? Could you move this there and test it that way instead?